pkg/resources/management.cattle.io/v3/projectroletemplatebinding/validator.go

```diff
@@ -141,50 +145,60 @@ func clusterFromProject(project string) (string, string) {
+		return field.Forbidden(fieldPath.Child("projectName"), "")
 	case oldPRTB.UserName != newPRTB.UserName && oldPRTB.UserName != "":
-		invalidFieldName = "userName"
+		return field.Forbidden(fieldPath.Child("userName"), "")
 	case oldPRTB.UserPrincipalName != newPRTB.UserPrincipalName && oldPRTB.UserPrincipalName != "":
-		invalidFieldName = "userPrincipalName"
+		return field.Forbidden(fieldPath.Child("userPrincipalName"), "")
 	case oldPRTB.GroupName != newPRTB.GroupName && oldPRTB.GroupName != "":
-		invalidFieldName = "groupName"
+		return field.Forbidden(fieldPath.Child("groupName"), "")
 	case oldPRTB.GroupPrincipalName != newPRTB.GroupPrincipalName && oldPRTB.GroupPrincipalName != "":
-		invalidFieldName = "groupPrincipalName"
+		return field.Forbidden(fieldPath.Child("groupPrincipalName"), "")
 	case (newPRTB.GroupName != "" || oldPRTB.GroupPrincipalName != "") && (newPRTB.UserName != "" || oldPRTB.UserPrincipalName != ""):
-		invalidFieldName = "both user and group"
+		return field.Forbidden(fieldPath.Child("groupName", "groupPrincipalName", "userName", "userPrincipalName"), "")
+	case newPRTB.ServiceAccount != "":
+		return field.Forbidden(fieldPath.Child("serviceAccount"), "")
 	default:
 		return nil
 	}
-
-	return fmt.Errorf("cannot update %s for clusterRoleTemplateBinding %s: %w", invalidFieldName, oldPRTB.Name, admission.ErrInvalidRequest)
 }
 
 // validateCreateFields checks if all required fields are present and valid.
-func (v *Validator) validateCreateFields(newPRTB *apisv3.ProjectRoleTemplateBinding) error {
+func (v *Validator) validateCreateFields(newPRTB *apisv3.ProjectRoleTemplateBinding, fieldPath *field.Path) *field.Error {
+	if newPRTB.ServiceAccount != "" {
+		return field.Forbidden(fieldPath.Child("serviceAccount"), "")
+	}
+
 	hasUserTarget := newPRTB.UserName != "" || newPRTB.UserPrincipalName != ""
 	hasGroupTarget := newPRTB.GroupName != "" || newPRTB.GroupPrincipalName != ""
 
 	if (hasUserTarget && hasGroupTarget) || (!hasUserTarget && !hasGroupTarget) {
-		return fmt.Errorf("binding must target either a user [userId]/[userPrincipalId] OR a group [groupId]/[groupPrincipalId]: %w", admission.ErrInvalidRequest)
+		return field.Invalid(fieldPath.Child("groupName", "groupPrincipalName", "userName", "userPrincipalName"),
+			"combination of [userId]/[userPrincipalId] and [groupId]/[groupPrincipalId]",
+			"binding must target either a user [userId]/[userPrincipalId] OR a group [groupId]/[groupPrincipalId]")
 	}
 
 	if newPRTB.ProjectName == "" {
-		return fmt.Errorf("binding must have field projectName set: %w", admission.ErrInvalidRequest)
+		return field.Required(fieldPath.Child("projectName"), "")
 	}
 
 	roleTemplate, err := v.roleTemplateResolver.RoleTemplateCache().Get(newPRTB.RoleTemplateName)
 	if err != nil {
-		return fmt.Errorf("unknown reference roleTemplate '%s': %w", newPRTB.RoleTemplateName, err)
+		return field.Invalid(fieldPath.Child("roleTemplate"), roleTemplate, "")
```

Comment:
I was not handling this error case correctly in my first iteration. I think we should modify the function to return an error if the error is anything other than `ErrorNotFound`. Then, the returning function can use `errors.Is(...)` to determine if it is a `field.Error` that should be sent back in a denied response or an error in the handler.

Reply:
What if we keep the logic and add the error text as the detail:
`return field.Invalid(fieldPath.Child("roleTemplate"), roleTemplate, err.Error())`

Reply:
This would still be undesired because we want the Admit function to return an error without a response object.